CLUSTER.md

```diff
@@ -0,0 +1,451 @@
+counter maintained by the Failover Coordinator. Every time a heartbeat is received by the Failover Coordinator, it is saved internally with the current tick
+(which is increased by the Failover Coordinator every K milliseconds). Since a node knows what tick its heartbeat was received in, it can tell how
+many ticks have passed since the primary sent its last heartbeat.
+
+### Resharding
+
+When a Topology Director receives an update which moves a set of slot to a specific shard it sets the appropriate slots to the
+importing/migrating state and bumps the topology epoch. This new slot topology will eventually reach all the relevant nodes.  
+
+It is expected that we will have atomic slot migration (based on fork+serialize like Redis replication) in Redis 
+_before_ Flotilla is GA - in which case Flotilla will only support that and not the existing key migration flow.
+Failing that, the behavior around key migration will be exactly same as in the current Redis Cluster - A node with a slot 
+set to Migrating returns ASK as appropriate, a node with a slot set to Importing requires Asking (and that behavior can
+be enhanced with things like bulk/full slot migrate without any impact on the Flotilla design).
+
+In our reference implementation we currently require the administrator to explicitly call the Topology Director with another update to terminate
+the Migrating/Importing state, but an alternative would be to handle this inside Flotilla by having nodes with migrating
+slots add the number of keys per slot to their heartbeat -- once that number drops to zero the Topology Director can automatically clear
+the state (It's also possible to emit the MIGRATE_DONE command from the primary migrating shard).
+
+![Resharding](images/resharding.jpg?raw=true "Failover")
+
+## Flotilla Implementation Details
+
+<!--
+Maybe mention Flotilla as a reference architecture and different possible implementations and an existing prototype implementation.
+-->
+
+### Roles
+
+While the Flotilla spec mostly discusses data nodes (and not Failover Coordinator & Topology Director implementation), in our reference implementation
+nodes within a Flotilla deployment have one or more roles:
+- Topology Director
+- Failover Coordinator
+- Data node
+
+The role of the node is specified in the join request (see [TBD Topics](#tbd-topics)).
+
+### Node and Failover Coordinator IDs
+
+All IDs in the system are:
+- Unique numbers
+- Monotonically increasing
+- Non-repeating
+
+The advantage to using numeric IDs is that:
+
+- It minimizes the payload for topology messages
+  (even if someone constantly adds nodes at one per <b>n</b>s(!!), it will ~500 years for the ID counters to exceed 64 bits).
+- It's easier to parse & communicate for humans.
+
+The disadvantage is it requires some garbage collection by the orchestrating admin when a node joining fails at the last second
```

Comment:
>it requires some garbage collection

Do we need to GC in the node removal case as well? Is the requirement for GC from the 64bit namespace? If so, can we consider 128bit instead so that we don't need any GC logic?

Reply:
No, the need to GC (as a manual action) could only occurs if a node fails to join in a specific manner.

Node -> TD request goes thorough
TD creates the relevant entity metadata
Node restarts / Network dies before response is received.

The need to GC in this case is to keep the node list from becoming full of dead nodes, not to prevent overflow.

worth noting that we will support join requests with an externally provided ID (in which case the TD just validates it's not being used), in this case one case safely resend the same join request if the above failure happens and if all the relevant metadata is the same + we have never received a heartbeat from the joining node the TD will approve it.

Reply:
>The need to GC in this case is to keep the node list from becoming full of dead nodes,

Got it. I read line 376 differently. This seems to be more implementation oriented IMO. I can also see other mechanisms like timeout would work. I understand your team is working on a reference implementation and would like to spec out the details as well but I think it would help a lot if we can review the two separately, i.e., a high-level spec of `what and why` and a detailed spec of `how`.

Reply:
This is detailed under the "Implementation details" section - as the design of the reference implementation evolves we might decide to split the two docs.